File: redis-iam-auth/client-example.js
/**
 * Example Node.js client for connecting to ElastiCache Redis using IAM authentication
 */

const AWS = require('aws-sdk');
const Redis = require('ioredis');

// Configure AWS SDK
AWS.config.update({ region: process.env.AWS_REGION || 'us-east-1' });

// Redis connection parameters
const REDIS_HOST = process.env.REDIS_HOST;
const REDIS_PORT = parseInt(process.env.REDIS_PORT || '6379');
const REDIS_CLUSTER_ID = process.env.REDIS_CLUSTER_ID;
const REDIS_USER = process.env.REDIS_USER || 'IAMAuthUser';

/**
 * Generate an authentication token using IAM credentials
 * @returns {Promise<string>} The authentication token
 */
async function getAuthToken() {
  try {
    const elasticache = new AWS.ElastiCache();
    const params = {
      ReplicationGroupId: REDIS_CLUSTER_ID,
      UserName: REDIS_USER
    };
    
    const data = await elasticache.generateAuthToken(params).promise();
    console.log('Successfully generated auth token');
    return data.AuthToken;
  } catch (error) {
    console.error('Failed to generate auth token:', error);
    throw error;
  }
}

/**
 * Connect to Redis using IAM authentication
 * @returns {Promise<Redis>} Redis client
 */
async function connectToRedis() {
  try {
    // Get authentication token
    const authToken = await getAuthToken();
    
    // Connect to Redis
    const client = new Redis({
      host: REDIS_HOST,
      port: REDIS_PORT,
      password: authToken,
      tls: true
    });
    
    // Handle connection events
    client.on('connect', () => {
      console.log('Connected to Redis');
    });
    
    client.on('error', (err) => {
      console.error('Redis connection error:', err);
    });
    
    // Test connection
    const pingResult = await client.ping();
    console.log(`Redis connection test: ${pingResult}`);
    
    return client;
  } catch (error) {
    console.error('Failed to connect to Redis:', error);
    throw error;
  }
}

/**
 * Main function to demonstrate Redis IAM authentication
 */
async function main() {
  try {
    // Check required environment variables
    if (!REDIS_HOST || !REDIS_CLUSTER_ID) {
      console.error('Required environment variables not set');
      console.error('Please set REDIS_HOST and REDIS_CLUSTER_ID');
      return;
    }
    
    // Connect to Redis
    const redisClient = await connectToRedis();
    
    // Example operations
    await redisClient.set('test_key', 'Hello from IAM auth!');
    const value = await redisClient.get('test_key');
    console.log(`Retrieved value: ${value}`);
    
    // Close connection
    await redisClient.quit();
  } catch (error) {
    console.error('Error in main function:', error);
  }
}

// Run the main function
main();

File: redis-iam-auth/client-example.py
#!/usr/bin/env python3
"""
Example client for connecting to ElastiCache Redis using IAM authentication
"""

import boto3
import redis
import os
import logging

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

# Redis connection parameters
REDIS_HOST = os.environ.get('REDIS_HOST')
REDIS_PORT = int(os.environ.get('REDIS_PORT', '6379'))
REDIS_CLUSTER_ID = os.environ.get('REDIS_CLUSTER_ID')
REDIS_USER = os.environ.get('REDIS_USER', 'IAMAuthUser')

def get_auth_token():
    """Generate an authentication token using IAM credentials"""
    try:
        elasticache = boto3.client('elasticache')
        auth_token = elasticache.generate_auth_token(
            ReplicationGroupId=REDIS_CLUSTER_ID,
            UserName=REDIS_USER
        )
        logger.info("Successfully generated auth token")
        return auth_token
    except Exception as e:
        logger.error(f"Failed to generate auth token: {e}")
        raise

def connect_to_redis():
    """Connect to Redis using IAM authentication"""
    try:
        # Get authentication token
        auth_token = get_auth_token()
        
        # Connect to Redis
        client = redis.Redis(
            host=REDIS_HOST,
            port=REDIS_PORT,
            password=auth_token,
            ssl=True,
            decode_responses=True
        )
        
        # Test connection
        response = client.ping()
        logger.info(f"Redis connection test: {response}")
        
        return client
    except Exception as e:
        logger.error(f"Failed to connect to Redis: {e}")
        raise

def main():
    """Main function to demonstrate Redis IAM authentication"""
    try:
        # Check required environment variables
        if not REDIS_HOST or not REDIS_CLUSTER_ID:
            logger.error("Required environment variables not set")
            logger.error("Please set REDIS_HOST and REDIS_CLUSTER_ID")
            return
        
        # Connect to Redis
        redis_client = connect_to_redis()
        
        # Example operations
        redis_client.set('test_key', 'Hello from IAM auth!')
        value = redis_client.get('test_key')
        logger.info(f"Retrieved value: {value}")
        
    except Exception as e:
        logger.error(f"Error in main function: {e}")

if __name__ == "__main__":
    main()
